src/circuit/gadget/sinsemilla/note_commit.rs

```diff
@@ -0,0 +1,1579 @@
+use group::GroupEncoding;
+use halo2::{
+    circuit::Layouter,
+    plonk::{Advice, Column, ConstraintSystem, Error, Expression, Selector},
+    poly::Rotation,
+};
+use pasta_curves::{
+    arithmetic::{CurveAffine, FieldExt},
+    pallas,
+};
+
+use crate::{
+    circuit::gadget::{
+        ecc::{
+            chip::{EccChip, EccPoint},
+            Point,
+        },
+        utilities::{bitrange_subset, bool_check, copy, CellValue, Var},
+    },
+    constants::T_P,
+};
+
+use super::{
+    chip::{SinsemillaChip, SinsemillaCommitDomains, SinsemillaConfig},
+    CommitDomain, Message, MessagePiece,
+};
+
+/*
+    <https://zips.z.cash/protocol/nu5.pdf#concretesinsemillacommit>
+    We need to hash g★_d || pk★_d || i2lebsp_{64}(v) || rho || psi,
+    where
+        - g★_d is the representation of the point g_d, with 255 bits used for the
+          x-coordinate and 1 bit used for the y-coordinate;
+        - pk★_d is the representation of the point pk_d, with 255 bits used for the
+          x-coordinate and 1 bit used for the y-coordinate;
+        - v is a 64-bit value;
+        - rho is a base field element (255 bits); and
+        - psi is a base field element (255 bits).
+*/
+
+#[allow(non_snake_case)]
+#[derive(Clone, Debug)]
+pub struct NoteCommitConfig {
+    q_canon_1: Selector,
+    q_canon_2: Selector,
+    q_canon_3: Selector,
+    advices: [Column<Advice>; 10],
+    sinsemilla_config: SinsemillaConfig,
+}
+
+impl NoteCommitConfig {
+    #[allow(non_snake_case)]
+    #[allow(clippy::many_single_char_names)]
+    pub(in crate::circuit) fn configure(
+        meta: &mut ConstraintSystem<pallas::Base>,
+        advices: [Column<Advice>; 10],
+        sinsemilla_config: SinsemillaConfig,
+    ) -> Self {
+        let q_canon_1 = meta.selector();
+        let q_canon_2 = meta.selector();
+        let q_canon_3 = meta.selector();
+
+        let config = Self {
+            q_canon_1,
+            q_canon_2,
+            q_canon_3,
+            advices,
+            sinsemilla_config,
+        };
+
+        // Useful constants
+        let two = pallas::Base::from_u64(2);
+        let two_pow_2 = pallas::Base::from_u64(1 << 2);
+        let two_pow_4 = two_pow_2.square();
+        let two_pow_5 = two_pow_4 * two;
+        let two_pow_6 = two_pow_5 * two;
+        let two_pow_8 = two_pow_4.square();
+        let two_pow_9 = two_pow_8 * two;
+        let two_pow_10 = two_pow_9 * two;
+        let two_pow_254 = pallas::Base::from_u128(1 << 127).square();
+        let two_pow_130 = Expression::Constant(pallas::Base::from_u128(1 << 65).square());
+        let two_pow_140 = Expression::Constant(pallas::Base::from_u128(1 << 70).square());
+        let t_p = Expression::Constant(pallas::Base::from_u128(T_P));
+
+        // Check that y = lsb + 2 * y_hi.
+        meta.create_gate("y decomposition check", |meta| {
```

Comment:
I'm pretty sure we also need to check the canonicity here, or else someone could witness a non-canonical `y_hi` in order to flip the `lsb` (since `p` is odd, adding it to a canonical scalar will result in an equivalent but non-canonical scalar with the opposite parity).